### PR TITLE
Upgrade path with SAD cases and code refactor

### PR DESCRIPTION
**Pull request by vaibhavhd**
### Description of PR

WAIT FOR MERGE of https://github.com/Azure/sonic-mgmt/pull/4876/files first

Summary: New SAD case upgrade path scenario and code reuse between upgrade and reboot path cases.
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
For better coverage of upgrade path scenario.

Changes added here:
1. New testcases - test_warm_upgrade_sad_path - supported for both Physical and KVM platform.
2. Enable log_analyzer for upgrade path scenario.
3.  Refactor upgrade path code and resuse the code from advanced_reboot.py - this reuse of code allows singular changes (not needed twice), and less code to maintain.

#### How did you do it?
Add SAD case for upgrade path using changes done in PR #4876.
Add log_analyzer fixture.
Remvoed duplicated code between upgrade_helpers and advanced_reboot.py

#### How did you verify/test it?
Tested on KVM platform.

### CODE REVIEW

File: tests/upgrade_path/test_upgrade_path.py [hunk 1]
```diff
@@ -80,25 +64,52 @@ def test_upgrade_path(localhost, duthosts, rand_one_dut_hostname, ptfhost, upgra
...
             logger.info("Upgrading to {}".format(to_image))
-            target_version = install_sonic(duthost, to_image, tbinfo)
-            test_params = ptf_params
-            test_params['target_version'] = target_version
-            test_params['reboot_type'] = get_reboot_command(duthost, upgrade_type)
-            prepare_testbed_ssh_keys(duthost, ptfhost, test_params['dut_username'])
-            log_file = "/tmp/advanced-reboot.ReloadTest.{}.log".format(datetime.now().strftime('%Y-%m-%d-%H:%M:%S'))
-            if test_params['reboot_type'] == reboot_ctrl_dict.get(REBOOT_TYPE_COLD).get("command"):
+            install_sonic(duthost, to_image, tbinfo)
+            if upgrade_type == REBOOT_TYPE_COLD: #reboot_ctrl_dict.get(REBOOT_TYPE_COLD).get("command"):
```
Comment (shi-su) [inline tests/upgrade_path/test_upgrade_path.py:68]: What is the purpose of leaving the comment here?

File: tests/upgrade_path/test_upgrade_path.py [hunk 2]
```diff
@@ -80,25 +64,52 @@ def test_upgrade_path(localhost, duthosts, rand_one_dut_hostname, ptfhost, upgra
...
+                advanceboot_loganalyzer=advanceboot_loganalyzer)
+            sad_preboot_list, sad_inboot_list = get_sad_case_list(duthost, nbrhosts, fanouthosts, tbinfo, sad_case_type)
+            advancedReboot.runRebootTestcase(
+                prebootList=sad_preboot_list,
+                inbootList=sad_inboot_list
+            )
+            reboot_cause = get_reboot_cause(duthost)
+            logger.info("Check reboot cause. Expected cause {}".format(upgrade_type))
+            pytest_assert(reboot_cause == upgrade_type, "Reboot cause {} did not match the trigger - {}".format(reboot_cause, upgrade_type))
+            check_services(duthost)
```
Comment (shi-su) [inline tests/upgrade_path/test_upgrade_path.py:115]: Maybe add a new line after this to fix its complaint?

File: tests/upgrade_path/test_upgrade_path.py (hunk 1 above)
Comment (vaibhavhd) [reply tests/upgrade_path/test_upgrade_path.py]: This was unintentional. Removed the comment now.

**Review by shi-su — APPROVED** [review]

**Pull request merged**